rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+---
+authors: Krzysztof Skrzętnicki (krzysztof.skrzetnicki@goteleport.com)
+state: draft
+---
+
+# RFD 151 - Database Permission Management
+
+## Required Approvals
+
+- Engineering: @r0mant && @smallinsky
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+The Database Permission Management is a proposal to introduce a new permission management system for databases integrated with Teleport. It aims to provide a structured and secure way to manage access permissions for users connecting to databases using Database Access.
+
+## Why
+
+As organizations increasingly rely on Teleport to access and manage their databases, there is a growing need for a unified and robust permission management system. The Database Permission Management feature will simplify the process of granting, revoking, and auditing database access permissions, ensuring that security and compliance requirements are met.
+
+In the most common mode of operation, Database Access connects the user to the database as a requested user, but does not verify if the user exists or has the appropriate rights. This task is left for the database administrator to perform. Automatic user creation is a relatively new Database Access feature that automates part of this task in supported database configurations. On connection, a new database user is created (or an old one reactivated) with a name matching that of the Teleport user. This new user is given roles from a fixed list of roles. Teleport expects the roles to be managed manually and has no insight into their definition.
+
+This RFD introduces a mechanism for _database permission management_, a way to define the effective database user permissions within Teleport roles and automatically apply them as needed. This provides visibility into the effective permissions and eases the burden on database administrators.
+
+## Details
+
+### UX
+
+#### Usage
+
+Teleport administrators modify the role definition assigned to the user, configuring both automatic user creation and individual permissions for the database.
+
+On connection, the engine calculates the effective permissions from the roles and database schema.
```

Comment:
How the db schema is fetched  on each connection/ cached ?. I wonder about connection latency in case of fetching all database schema during handling a new connection. 

Does auto user provisoning needs to be supported/configured by database to fetch db schema or we will use a sperate mechanism ?

Reply:
I can see how with huge, complex schemas the querying part can be a significant fraction of a second. I think this is fast enough to be acceptable in practice, as database connections are considered a heavy resource, and long-lived. For smaller databases, this should be near-instant, so even less of a problem. 

We can cache the database schema in the db agent, but we can easily add this later if it ever becomes an actual problem.

As for auto user provisioning, it is a hard dependency in this design. This means we already need a highly privileged user to configure new users, and may as well use it to query the database schema.

Reply:
I've ran some tests against Postgres. For a believable high-end case (1,000 tables, 20 columns each), querying the names of all columns took 34 ms, which is vanishingly small. Increasing the table factor by 10 for a total of 10,000 tables, the same query runs in 271 ms. I think it is unlikely there are actual databases with this many tables, but even then this is a reasonable amount of time.